Userland/Libraries/LibGfx/QOIWriter.cpp

```diff
@@ -0,0 +1,202 @@
+                int blue_difference = pixel.blue() - previous_pixel.blue();
+                int relative_red_difference = red_difference - green_difference;
+                int relative_blue_difference = blue_difference - green_difference;
+
+                if (red_difference > -3 && red_difference < 2
+                    && green_difference > -3 && green_difference < 2
+                    && blue_difference > -3 && blue_difference < 2) {
+                    writer.add_diff_chunk(red_difference, green_difference, blue_difference);
+                    previous_pixel = pixel;
+                    continue;
+                }
+                if (relative_red_difference > -9 && relative_red_difference < 8
+                    && green_difference > -33 && green_difference < 32
+                    && relative_blue_difference > -9 && relative_blue_difference < 8) {
+                    writer.add_luma_chunk(relative_red_difference, green_difference, relative_blue_difference);
+                    previous_pixel = pixel;
+                    continue;
+                }
+
+                writer.add_rgb_chunk(pixel.red(), pixel.green(), pixel.blue());
+                previous_pixel = pixel;
+                continue;
+            }
+
+            previous_pixel = pixel;
+
+            // Write full color values
+            writer.add_rgba_chunk(pixel.red(), pixel.green(), pixel.blue(), pixel.alpha());
+        }
+    }
+
+    writer.add_end_marker();
+
+    return ByteBuffer::copy(writer.m_data).release_value_but_fixme_should_propagate_errors();
+}
+
+void QOIWriter::add_header(u32 width, u32 height, Channels channels = Channels::RGBA, Colorspace color_space = Colorspace::sRGB)
+{
+    // TODO: Handle RGB and all linear channels
+    if (channels == Channels::RGB || color_space == Colorspace::LINEAR)
+        TODO();
+
+    m_data.append(qoi_magic_bytes.data(), sizeof(qoi_magic_bytes));
+
+    auto big_endian_width = AK::convert_between_host_and_big_endian(width);
+    m_data.append(bit_cast<u8*>(&big_endian_width), sizeof(width));
+
+    auto big_endian_height = AK::convert_between_host_and_big_endian(height);
+    m_data.append(bit_cast<u8*>(&big_endian_height), sizeof(height));
+
+    // Number of channels: 3 = RGB, 4 = RGBA
+    m_data.append(4);
+
+    // Colorspace: 0 = sRGB, 1 = all linear channels
+    m_data.append(color_space == Colorspace::sRGB ? 0 : 1);
+}
+
+void QOIWriter::add_rgb_chunk(u8 r, u8 g, u8 b)
+{
+    constexpr static u8 rgb_tag = 0b1111'1110;
+
+    m_data.append(rgb_tag);
+    m_data.append(r);
+    m_data.append(g);
+    m_data.append(b);
+}
+
+void QOIWriter::add_rgba_chunk(u8 r, u8 g, u8 b, u8 a)
+{
+    constexpr static u8 rgba_tag = 0b1111'1111;
+
+    m_data.append(rgba_tag);
+    m_data.append(r);
+    m_data.append(g);
+    m_data.append(b);
+    m_data.append(a);
+}
+
+void QOIWriter::add_index_chunk(unsigned int index)
+{
+    constexpr static u8 index_tag = 0b00 << 6;
```

Comment:
`0 << 6` is `0`. Mistake?

Reply:
Seeing the ones below - probably not. Might be clearer when writing out the full bits everywhere instead though.